`extend api` (#3802)
carbon-language/carbon-lang: proposals/p3802.md

[#2745](https://github.com/carbon-language/carbon-lang/issues/2745). This means
`Extending` is equivalent to:

```carbon
class Extending {
  fn G[self: Self]();
  alias F = Extended.F;
  impl as ImplicitAs(Extended);
}
```

The extended type is required to be complete at the point of the `extend api`
declaration, so the lookups that will be performed in the the containing class
definition into the extended type can succeed.

## Details

Normally simple member access on a value
[looks in the type of that value](/docs/design/expressions/member_access.md#values).
However, classes can designate other entities to also look up in using the
`extend` keyword (see [proposal #2760](/proposals/p2760.md#proposal)):

-   `extend impl as I` adds lookup into the implementation of an interface `I`.
-   `extend base: B` adds lookup into the base class `B`.
-   `extend adapt C` adds lookup into the adapted class `C`.
-   [mixins are also planning](/proposals/p2760.md#future-work-mixins) on using
    an `extend` syntax

Interfaces and named constraints also
[use the `extend` keyword](/docs/design/generics/details.md#interface-extension)

> to lookup into another, which is also required.

Reviewer: The antecedent of "which" here was unclear to me. I first read this as the lookup into another constraint being required, and then as the `extend` keyword being required (unlike in `extend impl`, where it can be omitted to turn off the `extend api` behavior). But I think this is "required" in the sense of a requirement -- implementing the other constraint is a requirement of the enclosing constraint.

Maybe:
```suggestion
to indicate that lookups should be performed in another constraint,
and to add a requirement that `Self` implements the other constraint.
```

> For some use cases, we want to be able to add name to lookup into that type

Reviewer: ```suggestion
For some use cases, we want to be able to add to name lookup into a type
```

without making other changes to the type, for which we introduce `extend api`,
with consistent lookup rules as the other uses of `extend`, with conflicts
handled as determined by leads issues
[#2355](https://github.com/carbon-language/carbon-lang/issues/2355) and
[#2745](https://github.com/carbon-language/carbon-lang/issues/2745).

The extended type is required to be complete at the point of the `extend api`
declaration, so these lookups into `T` can succeed.

The general rule for resolving ambiguity for `extend`, which we apply here as

> well, is that if lookup into `Box(T)` succeeds, then that result is used and no

Reviewer: This would be easier to follow if it appeared after the `Box` example below or had a link to that example.

Author: I rewrote it a bit, but let me know if it would be better to move this until after the `Box` example.

lookup into `T` is performed. If a class uses `extend` more than once, and finds
the same name more than once, that is an ambiguity error that needs to be
resolved by qualifying the name on use.

> **Future work:** Are these the right rules in the context of API evolution?
> See
> [comment on #3720](https://github.com/carbon-language/carbon-lang/pull/3720/files/97004ac6748a4f39e664d7409401df1811adc25c#r1527084183).

> **Note:** This is an alternative to defining an `alias` for each member of the
> extended type. This avoids having to repeat them, which is both lengthy and
> could get out of sync as the class evolves. The `extend api` approach works
> even if, as in this example, we don't know the names of the members of the
> type being extended.

Like other uses of `extend`, an `extend api` declaration do not have access

> control modifiers and only operate on public names.

Reviewer: ```suggestion
Like other uses of `extend`, an `extend api` declaration does not have access
control modifiers and only operates on public names.
```

### Use case: Member forwarding

Consider a class that we want to act like it has the members of another type.
For example, a type `Box(T)` that has a pointer to a `T` object allocated on the
heap:

```carbon
class Box(T:! type) {
  var ptr: T*;
  // ???
}
```

`Box(T)` should act like it has all the members of `T`:

```carbon
class String {
  fn Search[self: Self](c: u8) -> i32;
}
var b: Box(String) = ...;
var position: i32 = b.Search(32);
```

There are two ingredients to make this work:

-   We need some way to make `b.Search` be equivalent to `b.(String.Search)` not
    `b.(Box(String).Search)`.

> -   We need the act of binding a method of `String` to a `Box(String)` value

Reviewer: ```suggestion
-   We need the act of binding a method of `String` to a `Box(String)` value to
```

-   We have to be a little careful in the implementation of `ValueBind(Box(T))`
    to still use value binding even when we get a reference expression from
    dereferencing the pointer `ptr`.

With these two ingredients, `b.Search(32)` is equivalent to
`b.(String.Search)(32)` due to `extend api`, which then uses the custom binding
implementation to get something equivalent to `b.ptr->(String.Search)(32)`,
which is the same as `b.ptr->Search(32)` since `b.ptr` has type `String*`.

### Use case: Reference library type

The `Box` type that was used as
[an example of member forwarding](#use-case-member-forwarding) preserves the
expression category -- you can only call mutating operations on a `Box`
reference expression, not on a `Box` value expression. For a library reference
type that would simulate the C++-built-in reference type, we instead want to
permit value binding for anything where `T` permits reference binding. This can
also be done using `extend api` and a different set of binding implementations:

```carbon
class Ref(T:! type) {
  var ptr: T*;
  extend api T;
}
impl forall [T:! type, U:! RefBind(T)]
    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {

>   fn Op[self: Self](p: Ref(T)) -> Ref(Result) {

Reviewer: ```suggestion
  fn Op[self: Self](p: Ref(T)) -> Ref(U.Result) {
```
or
```suggestion
  fn Op[self: Self](p: Ref(T)) -> Result {
```
I think `Result` here names `(U as ValueBind(Ref(T))).Result`, which is `Ref(U.Result)`.

>     return {.ptr = self.Op(p->ptr)};

Reviewer: ```suggestion
    return {.ptr = self.Op(p.ptr)};
```

  }
}

> // (Plus impls for the other three combinations of ValueBind and RefBind.)

Reviewer: Exploring these myself:

```carbon
// <Ref(T) value>.<ref bindable> produces a Ref(R) value expression.
impl forall [T:! type, U:! RefBind(T)]
    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return {.ptr = self.Op(p->ptr)};
  }
}

// <Ref(T) ref expr>.<ref bindable> produces a reference expression.
// Reference collapse: no added Ref(...) here.
impl forall [T:! type, U:! RefBind(T)]
    U as RefBind(Ref(T)) where .Result = U.Result {
  fn Op[self: Self](p: Ref(T)*) -> Result* {
    return self.Op(p->ptr);
  }
}

// <Ref(T) value>.<value bindable> produces a value expression.
// A value binding is performed to turn the Ref(T) into a T.
impl forall [T:! type, U:! ValueBind(T)]
    U as ValueBind(Ref(T)) where .Result = U.Result {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return self.Op(*p.ptr);
  }
}
```

I think implementing `U as RefBind(Ref(T))` in terms of `U as ValueBind(T)` is not possible. We can't produce a pointer as a result of the binding.

There's also overlap between the first and third `impl`s. And the first and second violate the new coherence rule in #3720:  given `ref: Ref({.n: i32})`, `ref.n` will either be a value expression of type `Ref(i32)` or a reference expression of type `i32` depending on the expression category of `ref`.

I think maybe the right thing to do is to drop the second `impl` entirely, and have only two `impl`s, both for `ValueBind`, with a `match_first`:

```carbon
match_first {

// <Ref(T) value>.<ref bindable> produces a value expression of type `Ref(R)`.
impl forall [T:! type, U:! RefBind(T)]
    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return {.ptr = self.Op(p->ptr)};
  }
}

// <Ref(T) value>.<value bindable> produces a value expression of type `R`.
// A value binding is performed to turn the Ref(T) into a T.
impl forall [T:! type, U:! ValueBind(T)]
    U as ValueBind(Ref(T)) where .Result = U.Result {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return self.Op(*p.ptr);
  }
}

}
```

For this to really work, I think we'd need binding to fall back from `RefBind` to `ValueBind` if reference binding isn't possible, as discussed in #3720. The above pair of `impl`s would then be implementing the same fallback semantics for `Ref(T)`.

Author: I don't think this works. #3720 only ever produces a reference expression (by automatically dereferencing the returned pointer) when the expression on the left side of the `.` is a reference expression. I think this makes `Ref(T)` impossible, and I should probably delete this section.

Author: Sorry, I had not understood what you were proposing (that you can't get a reference expression, but can get a `Ref(T)` value that acts like one). I've made these changes, added the additional support for calls on `Ref(T)` values, and moved this section to future work, [as discussed in open discussion on 05-16](https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.3qlpp5e56u46).

impl forall [S:! type, T:! Call(S)] Ref(T) as Call(S) where .Result = T.Result {
  // ...
}
```

For example, it could be used as:

```carbon
class C {
  fn F[addr self: Self*]();
  var n: i32;
}
fn F(r: Ref(C)) {
  // OK, even though `r` is a value expression.
  // `r.F` is a `Ref(<bound member function>)`,
  // which is callable.
  r.F();

  // OK, forms a reference to the `n` member.
  let n_ref: Ref(i32) = r.n;
}
```

Note, however, `Ref(T)` doesn't preserve the interface implementations of `T`,
see [future work](#future-work-customizing-impl-binding).

This application of `extend api` and binding was first described in
[this comment on #3720](https://github.com/carbon-language/carbon-lang/pull/3720/files/37e967ff53e89e345eecb49ec79c6dfbe18a3c54#r1513712572).

> ### Use case: with implicit conversion

Reviewer: Note to self: I'm parking this review here while looking at other things.
